@SergeyKanzhelev lint ci fails:
```
ERROR: staging/src/k8s.io/cri-client/pkg/remote_image_test.go:99:2: require-error: for error assertions use require (testifylint)
ERROR: 	assert.ErrorContains(t, err, "endpoint is required")
ERROR: 	^
```

--- a/cmd/kubeadm/app/util/runtime/impl.go
+++ b/cmd/kubeadm/app/util/runtime/impl.go
@@ -41,11 +41,17 @@ type Impl interface {
 }
 
 func (*defaultImpl) NewRemoteRuntimeService(ctx context.Context, endpoint string, connectionTimeout time.Duration) (criapi.RuntimeService, error) {
-	return criclient.NewRemoteRuntimeService(ctx, endpoint, connectionTimeout, nil, false)
+	return criclient.NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(connectionTimeout).
+		Build(ctx)
 }
 
 func (*defaultImpl) NewRemoteImageService(ctx context.Context, endpoint string, connectionTimeout time.Duration) (criapi.ImageManagerService, error) {
-	return criclient.NewRemoteImageService(ctx, endpoint, connectionTimeout, nil, false)
+	return criclient.NewRemoteImageServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(connectionTimeout).
+		Build(ctx)
 }
 
 func (*defaultImpl) RuntimeConfig(ctx context.Context, runtimeService criapi.RuntimeService) (*runtimeapi.RuntimeConfigResponse, error) {

--- a/cmd/kubemark/app/hollow_node.go
+++ b/cmd/kubemark/app/hollow_node.go
@@ -24,7 +24,6 @@ import (
 
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
-	"go.opentelemetry.io/otel/trace/noop"
 	internalapi "k8s.io/cri-api/pkg/apis"
 	"k8s.io/klog/v2"
 
@@ -247,14 +246,20 @@ func run(ctx context.Context, config *hollowNodeConfig) error {
 			return fmt.Errorf("Failed to start fake runtime, error: %w", err)
 		}
 		defer fakeRemoteRuntime.Stop()
-		runtimeService, err := remote.NewRemoteRuntimeService(ctx, endpoint, 15*time.Second, noop.NewTracerProvider(), false)
+		runtimeService, err := remote.NewRemoteRuntimeServiceBuilder().
+			WithEndpoint(endpoint).
+			WithConnectionTimeout(15 * time.Second).
+			Build(ctx)
 		if err != nil {
 			return fmt.Errorf("Failed to init runtime service, error: %w", err)
 		}
 
 		var imageService internalapi.ImageManagerService = fakeRemoteRuntime.ImageService
 		if config.UseHostImageService {
-			imageService, err = remote.NewRemoteImageService(ctx, c.ImageServiceEndpoint, 15*time.Second, noop.NewTracerProvider(), false)
+			imageService, err = remote.NewRemoteImageServiceBuilder().
+				WithEndpoint(c.ImageServiceEndpoint).
+				WithConnectionTimeout(15 * time.Second).
+				Build(ctx)
 			if err != nil {
 				return fmt.Errorf("Failed to init image service, error: %w", err)
 			}

--- a/pkg/kubelet/kubelet.go
+++ b/pkg/kubelet/kubelet.go
@@ -408,10 +408,20 @@ func PreInitRuntimeService(ctx context.Context, kubeCfg *kubeletconfiginternal.K
 	}
 	var err error
 	useStreaming := utilfeature.DefaultFeatureGate.Enabled(features.CRIListStreaming)
-	if kubeDeps.RemoteRuntimeService, err = remote.NewRemoteRuntimeService(ctx, kubeCfg.ContainerRuntimeEndpoint, kubeCfg.RuntimeRequestTimeout.Duration, kubeDeps.TracerProvider, useStreaming); err != nil {
+	if kubeDeps.RemoteRuntimeService, err = remote.NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(kubeCfg.ContainerRuntimeEndpoint).
+		WithConnectionTimeout(kubeCfg.RuntimeRequestTimeout.Duration).
+		WithTracerProvider(kubeDeps.TracerProvider).
+		WithUseStreaming(useStreaming).
+		Build(ctx); err != nil {
 		return err
 	}
-	if kubeDeps.RemoteImageService, err = remote.NewRemoteImageService(ctx, remoteImageEndpoint, kubeCfg.RuntimeRequestTimeout.Duration, kubeDeps.TracerProvider, useStreaming); err != nil {
+	if kubeDeps.RemoteImageService, err = remote.NewRemoteImageServiceBuilder().
+		WithEndpoint(remoteImageEndpoint).
+		WithConnectionTimeout(kubeCfg.RuntimeRequestTimeout.Duration).
+		WithTracerProvider(kubeDeps.TracerProvider).
+		WithUseStreaming(useStreaming).
+		Build(ctx); err != nil {
 		return err
 	}
 

--- a/pkg/kubelet/kubelet_test.go
+++ b/pkg/kubelet/kubelet_test.go
@@ -34,7 +34,6 @@ import (
 
 	sdktrace "go.opentelemetry.io/otel/sdk/trace"
 	"go.opentelemetry.io/otel/sdk/trace/tracetest"
-	oteltrace "go.opentelemetry.io/otel/trace"
 	noopoteltrace "go.opentelemetry.io/otel/trace/noop"
 
 	"k8s.io/component-base/metrics/legacyregistry"
@@ -65,7 +64,6 @@ import (
 	"k8s.io/component-base/metrics/testutil"
 	ndf "k8s.io/component-helpers/nodedeclaredfeatures"
 	ndftesting "k8s.io/component-helpers/nodedeclaredfeatures/testing"
-	internalapi "k8s.io/cri-api/pkg/apis"
 	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
 	remote "k8s.io/cri-client/pkg"
 	fakeremote "k8s.io/cri-client/pkg/fake"
@@ -3293,12 +3291,6 @@ func createAndStartFakeRemoteRuntime(t *testing.T) (*fakeremote.RemoteRuntime, s
 	return fakeRuntime, endpoint
 }
 
-func createRemoteRuntimeService(ctx context.Context, endpoint string, t *testing.T, tp oteltrace.TracerProvider) internalapi.RuntimeService {
-	runtimeService, err := remote.NewRemoteRuntimeService(ctx, endpoint, 15*time.Second, tp, false)
-	require.NoError(t, err)
-	return runtimeService
-}
-
 func TestNewMainKubeletStandAlone(t *testing.T) {
 	logger, tCtx := ktesting.NewTestContext(t)
 	tempDir, err := os.MkdirTemp("", "logs")
@@ -3367,7 +3359,11 @@ func TestNewMainKubeletStandAlone(t *testing.T) {
 		fakeRuntime.Stop()
 	}()
 	fakeRecorder := &record.FakeRecorder{}
-	rtSvc := createRemoteRuntimeService(tCtx, endpoint, t, noopoteltrace.NewTracerProvider())
+	rtSvc, err := remote.NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(15 * time.Second).
+		Build(tCtx)
+	require.NoError(t, err)
 	kubeDep := &Dependencies{
 		Auth:                 nil,
 		CAdvisorInterface:    cadvisor,
@@ -3526,7 +3522,11 @@ func TestNewMainKubeletWithCertAndCAReloadingEnabled(t *testing.T) {
 		fakeRuntime.Stop()
 	}()
 	fakeRecorder := &record.FakeRecorder{}
-	rtSvc := createRemoteRuntimeService(tCtx, endpoint, t, noopoteltrace.NewTracerProvider())
+	rtSvc, err := remote.NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(15 * time.Second).
+		Build(tCtx)
+	require.NoError(t, err)
 	kubeDep := &Dependencies{
 		Auth:                 nil,
 		CAdvisorInterface:    cadvisor,
@@ -3628,12 +3628,21 @@ func TestSyncPodSpans(t *testing.T) {
 	defer func() {
 		fakeRuntime.Stop()
 	}()
-	runtimeSvc := createRemoteRuntimeService(tCtx, endpoint, t, tp)
+	runtimeSvc, err := remote.NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(15 * time.Second).
+		WithTracerProvider(tp).
+		Build(tCtx)
+	require.NoError(t, err)
 	kubelet.runtimeService = runtimeSvc
 
 	fakeRuntime.ImageService.SetFakeImageSize(100)
 	fakeRuntime.ImageService.SetFakeImages([]string{"test:latest"})
-	imageSvc, err := remote.NewRemoteImageService(tCtx, endpoint, 15*time.Second, tp, false)
+	imageSvc, err := remote.NewRemoteImageServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(15 * time.Second).
+		WithTracerProvider(tp).
+		Build(tCtx)
 	assert.NoError(t, err)
 
 	kubelet.containerRuntime, _, err = kuberuntime.NewKubeGenericRuntimeManager(

--- a/staging/src/k8s.io/cri-client/pkg/remote_image.go
+++ b/staging/src/k8s.io/cri-client/pkg/remote_image.go
@@ -27,6 +27,7 @@ import (
 	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
 	"go.opentelemetry.io/otel/propagation"
 	"go.opentelemetry.io/otel/trace"
+	"go.opentelemetry.io/otel/trace/noop"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/backoff"
 	"google.golang.org/grpc/codes"
@@ -53,22 +54,83 @@ type remoteImageService struct {
 	useStreaming atomic.Bool
 }
 
-// NewRemoteImageService creates a new internalapi.ImageManagerService.
-// If useStreaming is true, streaming RPCs will be used for list operations
-// instead of unary RPCs. If the runtime returns an Unimplemented error,
-// the client automatically falls back to unary RPCs.
-// NOTE: useStreaming is supposed to be gated by the CRIListStreaming feature
-// gate and is expected to default to true once the feature graduates to GA,
-// at which point this parameter may be removed.
-func NewRemoteImageService(ctx context.Context, endpoint string, connectionTimeout time.Duration, tp trace.TracerProvider, useStreaming bool) (internalapi.ImageManagerService, error) {
+// RemoteImageServiceBuilder builds a new internalapi.ImageManagerService.
+//
+// Construct a builder with NewRemoteImageServiceBuilder, then chain With*
+// methods to set non-default options before calling Build. Adding new options
+// here is preferred over adding parameters to NewRemoteImageService so that
+// default values can be changed in a single place without having to update
+// every callsite.
+type RemoteImageServiceBuilder struct {
+	endpoint          string
+	connectionTimeout time.Duration
+	tracerProvider    trace.TracerProvider
+	// tracerProviderSet tracks whether WithTracerProvider was called, so that
+	// an explicit nil can be distinguished from the unset default and used to
+	// opt out of installing the otelgrpc stats handler entirely.
+	tracerProviderSet bool
+	// useStreaming indicates whether to use streaming RPCs for list operations
+	// when the CRIListStreaming feature gate is enabled. It is expected to
+	// default to true once the feature graduates to GA.
+	useStreaming bool
+}
+
+// NewRemoteImageServiceBuilder returns a builder with default options for
+// constructing a remote image service.
+func NewRemoteImageServiceBuilder() *RemoteImageServiceBuilder {
+	return &RemoteImageServiceBuilder{}
+}
+
+// WithEndpoint sets the gRPC endpoint of the remote image service.
+func (b *RemoteImageServiceBuilder) WithEndpoint(endpoint string) *RemoteImageServiceBuilder {
+	b.endpoint = endpoint
+	return b
+}
+
+// WithConnectionTimeout sets the timeout used when connecting to the remote
+// image service.
+func (b *RemoteImageServiceBuilder) WithConnectionTimeout(connectionTimeout time.Duration) *RemoteImageServiceBuilder {
+	b.connectionTimeout = connectionTimeout
+	return b
+}
+
+// WithTracerProvider sets the OpenTelemetry tracer provider used to
+// instrument the gRPC client. If WithTracerProvider is not called, the
+// otelgrpc stats handler is installed with a noop tracer provider, so no
+// traces are produced but gRPC context propagation still works. Passing a nil
+// provider explicitly opts out of installing the stats handler entirely,
+// disabling both tracing and context propagation.
+func (b *RemoteImageServiceBuilder) WithTracerProvider(tp trace.TracerProvider) *RemoteImageServiceBuilder {
+	b.tracerProvider = tp
+	b.tracerProviderSet = true
+	return b
+}
+
+// WithUseStreaming controls whether streaming RPCs are used for list
+// operations. If the runtime returns an Unimplemented error, the client
+// automatically falls back to the corresponding unary RPC.
+func (b *RemoteImageServiceBuilder) WithUseStreaming(useStreaming bool) *RemoteImageServiceBuilder {
+	b.useStreaming = useStreaming
+	return b
+}
+
+// Build creates a new internalapi.ImageManagerService using the configured
+// options.
+func (b *RemoteImageServiceBuilder) Build(ctx context.Context) (internalapi.ImageManagerService, error) {
+	if b.endpoint == "" {
+		return nil, errors.New("endpoint is required")
+	}
+	if b.connectionTimeout <= 0 {
+		return nil, errors.New("connectionTimeout must be positive")
+	}
 	logger := klog.FromContext(ctx)
-	logger.V(3).Info("Connecting to image service", "endpoint", endpoint)
-	addr, dialer, err := util.GetAddressAndDialer(endpoint)
+	logger.V(3).Info("Connecting to image service", "endpoint", b.endpoint)
+	addr, dialer, err := util.GetAddressAndDialer(b.endpoint)
 	if err != nil {
 		return nil, err
 	}
 
-	ctx, cancel := context.WithTimeout(ctx, connectionTimeout)
+	ctx, cancel := context.WithTimeout(ctx, b.connectionTimeout)
 	defer cancel()
 
 	var dialOpts []grpc.DialOption
@@ -77,14 +139,20 @@ func NewRemoteImageService(ctx context.Context, endpoint string, connectionTimeo
 		grpc.WithAuthority("localhost"),
 		grpc.WithContextDialer(dialer),
 		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMsgSize)))
-	if tp != nil {
+	// When no tracer provider was configured, fall back to a noop provider
+	// so context propagation still works without producing real traces.
+	// See https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/examples/passthrough
+	if !b.tracerProviderSet {
+		b.tracerProvider = noop.NewTracerProvider()
+	}
+	// Install the otelgrpc stats handler unless the caller explicitly opted
+	// out by calling WithTracerProvider(nil).
+	if b.tracerProvider != nil {
 		tracingOpts := []otelgrpc.Option{
 			otelgrpc.WithMessageEvents(otelgrpc.ReceivedEvents, otelgrpc.SentEvents),
 			otelgrpc.WithPropagators(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{})),
-			otelgrpc.WithTracerProvider(tp),
+			otelgrpc.WithTracerProvider(b.tracerProvider),
 		}
-		// Even if there is no TracerProvider, the otelgrpc still handles context propagation.
-		// See https://github.com/open-telemetry/opentelemetry-go/tree/main/example/passthrough
 		dialOpts = append(dialOpts,
 			grpc.WithStatsHandler(otelgrpc.NewClientHandler(tracingOpts...)))
 	}
@@ -106,17 +174,33 @@ func NewRemoteImageService(ctx context.Context, endpoint string, connectionTimeo
 	}
 
 	service := &remoteImageService{
-		timeout: connectionTimeout,
+		timeout: b.connectionTimeout,
 		conn:    conn,
 	}
-	service.useStreaming.Store(useStreaming)
+	service.useStreaming.Store(b.useStreaming)
 
-	if err := service.validateServiceConnection(ctx, conn, endpoint); err != nil {
+	if err := service.validateServiceConnection(ctx, conn, b.endpoint); err != nil {
 		return nil, fmt.Errorf("validate service connection: %w", err)
 	}
 
 	return service, nil
+}
 
+// NewRemoteImageService creates a new internalapi.ImageManagerService.
+// If useStreaming is true, streaming RPCs will be used for list operations
+// instead of unary RPCs. If the runtime returns an Unimplemented error,
+// the client automatically falls back to unary RPCs.
+//
+// Deprecated: Use NewRemoteImageServiceBuilder so that default values
+// (e.g. useStreaming) can be changed in a single place without updating
+// every callsite.
+func NewRemoteImageService(ctx context.Context, endpoint string, connectionTimeout time.Duration, tp trace.TracerProvider, useStreaming bool) (internalapi.ImageManagerService, error) {
+	return NewRemoteImageServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(connectionTimeout).
+		WithTracerProvider(tp).
+		WithUseStreaming(useStreaming).
+		Build(ctx)
 }
 
 // Close will shutdown the internal gRPC client connection.

--- a/staging/src/k8s.io/cri-client/pkg/remote_image_test.go
+++ b/staging/src/k8s.io/cri-client/pkg/remote_image_test.go
@@ -26,26 +26,9 @@ import (
 	"github.com/stretchr/testify/require"
 	sdktrace "go.opentelemetry.io/otel/sdk/trace"
 	"go.opentelemetry.io/otel/sdk/trace/tracetest"
-	oteltrace "go.opentelemetry.io/otel/trace"
-	"go.opentelemetry.io/otel/trace/noop"
-	internalapi "k8s.io/cri-api/pkg/apis"
 	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
 	"k8s.io/cri-client/pkg/util"
 )
-
-func createRemoteImageServiceWithTracerProvider(ctx context.Context, endpoint string, tp oteltrace.TracerProvider, t *testing.T) internalapi.ImageManagerService {
-	imageService, err := NewRemoteImageService(ctx, endpoint, defaultConnectionTimeout, tp, false)
-	require.NoError(t, err)
-
-	return imageService
-}
-
-func createRemoteImageServiceWithoutTracerProvider(ctx context.Context, endpoint string, t *testing.T) internalapi.ImageManagerService {
-	imageService, err := NewRemoteImageService(ctx, endpoint, defaultConnectionTimeout, noop.NewTracerProvider(), false)
-	require.NoError(t, err)
-
-	return imageService
-}
 
 func TestImageServiceSpansWithTP(t *testing.T) {
 	fakeRuntime, endpoint := createAndStartFakeRemoteRuntime(t)
@@ -63,7 +46,12 @@ func TestImageServiceSpansWithTP(t *testing.T) {
 		sdktrace.WithBatcher(exp),
 	)
 	ctx := context.Background()
-	imgSvc := createRemoteImageServiceWithTracerProvider(ctx, endpoint, tp, t)
+	imgSvc, err := NewRemoteImageServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(defaultConnectionTimeout).
+		WithTracerProvider(tp).
+		Build(ctx)
+	require.NoError(t, err)
 	imgRef, err := imgSvc.PullImage(ctx, &runtimeapi.ImageSpec{Image: "busybox"}, nil, nil)
 	assert.NoError(t, err)
 	assert.Equal(t, "busybox", imgRef)
@@ -89,7 +77,11 @@ func TestImageServiceSpansWithoutTP(t *testing.T) {
 		sdktrace.WithBatcher(exp),
 	)
 	ctx := context.Background()
-	imgSvc := createRemoteImageServiceWithoutTracerProvider(ctx, endpoint, t)
+	imgSvc, err := NewRemoteImageServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(defaultConnectionTimeout).
+		Build(ctx)
+	require.NoError(t, err)
 	imgRef, err := imgSvc.PullImage(ctx, &runtimeapi.ImageSpec{Image: "busybox"}, nil, nil)
 	assert.NoError(t, err)
 	assert.Equal(t, "busybox", imgRef)
@@ -97,6 +89,19 @@ func TestImageServiceSpansWithoutTP(t *testing.T) {
 	err = tp.ForceFlush(ctx)
 	require.NoError(t, err)
 	assert.Empty(t, exp.GetSpans())
+}
+
+func TestImageServiceBuildValidatesRequiredOptions(t *testing.T) {
+	ctx := context.Background()
+	_, err := NewRemoteImageServiceBuilder().
+		WithConnectionTimeout(defaultConnectionTimeout).
+		Build(ctx)
+	require.ErrorContains(t, err, "endpoint is required")
+
+	_, err = NewRemoteImageServiceBuilder().
+		WithEndpoint("unix:///tmp/cri-client-test.sock").
+		Build(ctx)
+	require.ErrorContains(t, err, "connectionTimeout must be positive")
 }
 
 func TestNewRemoteImageServiceUnixSocketEndpoint(t *testing.T) {
@@ -118,7 +123,11 @@ func TestNewRemoteImageServiceUnixSocketEndpoint(t *testing.T) {
 	}()
 
 	ctx := context.Background()
-	imgSvc := createRemoteImageServiceWithoutTracerProvider(ctx, endpoint, t)
+	imgSvc, err := NewRemoteImageServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(defaultConnectionTimeout).
+		Build(ctx)
+	require.NoError(t, err)
 	info, err := imgSvc.ImageFsInfo(ctx)
 	require.NoError(t, err)
 	assert.NotNil(t, info)

--- a/staging/src/k8s.io/cri-client/pkg/remote_runtime.go
+++ b/staging/src/k8s.io/cri-client/pkg/remote_runtime.go
@@ -28,6 +28,7 @@ import (
 	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
 	"go.opentelemetry.io/otel/propagation"
 	"go.opentelemetry.io/otel/trace"
+	"go.opentelemetry.io/otel/trace/noop"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/backoff"
 	"google.golang.org/grpc/codes"
@@ -85,21 +86,85 @@ const (
 	CRIVersionV1 CRIVersion = "v1"
 )
 
-// NewRemoteRuntimeService creates a new internalapi.RuntimeService.
-// If useStreaming is true, streaming RPCs will be used for list operations
-// instead of unary RPCs. If the runtime returns an Unimplemented error,
-// the client automatically falls back to unary RPCs.
+// RemoteRuntimeServiceBuilder builds a new internalapi.RuntimeService.
+//
+// Construct a builder with NewRemoteRuntimeServiceBuilder, then chain With*
+// methods to set non-default options before calling Build. Adding new options
+// here is preferred over adding parameters to NewRemoteRuntimeService so that
+// default values can be changed in a single place without having to update
+// every callsite.
+type RemoteRuntimeServiceBuilder struct {
+	endpoint          string
+	connectionTimeout time.Duration
+	tracerProvider    trace.TracerProvider
+	// tracerProviderSet tracks whether WithTracerProvider was called, so that
+	// an explicit nil can be distinguished from the unset default and used to
+	// opt out of installing the otelgrpc stats handler entirely.
+	tracerProviderSet bool
+	// useStreaming indicates whether to use streaming RPCs for list operations
+	// when the CRIListStreaming feature gate is enabled. It is expected to
+	// default to true once the feature graduates to GA.
+	useStreaming bool
+}
+
+// NewRemoteRuntimeServiceBuilder returns a builder with default options for
+// constructing a remote runtime service.
+func NewRemoteRuntimeServiceBuilder() *RemoteRuntimeServiceBuilder {
+	return &RemoteRuntimeServiceBuilder{}
+}
+
+// WithEndpoint sets the gRPC endpoint of the remote runtime service.
+func (b *RemoteRuntimeServiceBuilder) WithEndpoint(endpoint string) *RemoteRuntimeServiceBuilder {
+	b.endpoint = endpoint
+	return b
+}
+
+// WithConnectionTimeout sets the timeout used when connecting to the remote
+// runtime service.
+func (b *RemoteRuntimeServiceBuilder) WithConnectionTimeout(connectionTimeout time.Duration) *RemoteRuntimeServiceBuilder {
+	b.connectionTimeout = connectionTimeout
+	return b
+}
+
+// WithTracerProvider sets the OpenTelemetry tracer provider used to
+// instrument the gRPC client. If WithTracerProvider is not called, the
+// otelgrpc stats handler is installed with a noop tracer provider, so no
+// traces are produced but gRPC context propagation still works. Passing a nil
+// provider explicitly opts out of installing the stats handler entirely,
+// disabling both tracing and context propagation.
+func (b *RemoteRuntimeServiceBuilder) WithTracerProvider(tp trace.TracerProvider) *RemoteRuntimeServiceBuilder {
+	b.tracerProvider = tp
+	b.tracerProviderSet = true
+	return b
+}
+
+// WithUseStreaming controls whether streaming RPCs are used for list
+// operations. If the runtime returns an Unimplemented error, the client
+// automatically falls back to the corresponding unary RPC.
 // NOTE: useStreaming is supposed to be gated by the CRIListStreaming feature
 // gate and is expected to default to true once the feature graduates to GA,
-// at which point this parameter may be removed.
-func NewRemoteRuntimeService(ctx context.Context, endpoint string, connectionTimeout time.Duration, tp trace.TracerProvider, useStreaming bool) (internalapi.RuntimeService, error) {
+// at which point this method may be removed.
+func (b *RemoteRuntimeServiceBuilder) WithUseStreaming(useStreaming bool) *RemoteRuntimeServiceBuilder {
+	b.useStreaming = useStreaming
+	return b
+}
+
+// Build creates a new internalapi.RuntimeService using the configured
+// options.
+func (b *RemoteRuntimeServiceBuilder) Build(ctx context.Context) (internalapi.RuntimeService, error) {
+	if b.endpoint == "" {
+		return nil, errors.New("endpoint is required")
+	}
+	if b.connectionTimeout <= 0 {
+		return nil, errors.New("connectionTimeout must be positive")
+	}
 	logger := klog.FromContext(ctx)
-	logger.V(3).Info("Connecting to runtime service", "endpoint", endpoint)
-	addr, dialer, err := util.GetAddressAndDialer(endpoint)
+	logger.V(3).Info("Connecting to runtime service", "endpoint", b.endpoint)
+	addr, dialer, err := util.GetAddressAndDialer(b.endpoint)
 	if err != nil {
 		return nil, err
 	}
-	ctx, cancel := context.WithTimeout(ctx, connectionTimeout)
+	ctx, cancel := context.WithTimeout(ctx, b.connectionTimeout)
 	defer cancel()
 
 	var dialOpts []grpc.DialOption
@@ -108,14 +173,20 @@ func NewRemoteRuntimeService(ctx context.Context, endpoint string, connectionTim
 		grpc.WithAuthority("localhost"),
 		grpc.WithContextDialer(dialer),
 		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(maxMsgSize)))
-	if tp != nil {
+	// When no tracer provider was configured, fall back to a noop provider
+	// so context propagation still works without producing real traces.
+	// See https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/examples/passthrough
+	if !b.tracerProviderSet {
+		b.tracerProvider = noop.NewTracerProvider()
+	}
+	// Install the otelgrpc stats handler unless the caller explicitly opted
+	// out by calling WithTracerProvider(nil).
+	if b.tracerProvider != nil {
 		tracingOpts := []otelgrpc.Option{
 			otelgrpc.WithMessageEvents(otelgrpc.ReceivedEvents, otelgrpc.SentEvents),
 			otelgrpc.WithPropagators(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{})),
-			otelgrpc.WithTracerProvider(tp),
+			otelgrpc.WithTracerProvider(b.tracerProvider),
 		}
-		// Even if there is no TracerProvider, the otelgrpc still handles context propagation.
-		// See https://github.com/open-telemetry/opentelemetry-go/tree/main/example/passthrough
 		dialOpts = append(dialOpts,
 			grpc.WithStatsHandler(otelgrpc.NewClientHandler(tracingOpts...)))
 	}
@@ -137,17 +208,34 @@ func NewRemoteRuntimeService(ctx context.Context, endpoint string, connectionTim
 	}
 
 	service := &remoteRuntimeService{
-		timeout:      connectionTimeout,
+		timeout:      b.connectionTimeout,
 		logReduction: logreduction.NewLogReduction(identicalErrorDelay),
 		conn:         conn,
 	}
-	service.useStreaming.Store(useStreaming)
+	service.useStreaming.Store(b.useStreaming)
 
-	if err := service.validateServiceConnection(ctx, conn, endpoint); err != nil {
+	if err := service.validateServiceConnection(ctx, conn, b.endpoint); err != nil {
 		return nil, fmt.Errorf("validate service connection: %w", err)
 	}
 
 	return service, nil
+}
+
+// NewRemoteRuntimeService creates a new internalapi.RuntimeService.
+// If useStreaming is true, streaming RPCs will be used for list operations
+// instead of unary RPCs. If the runtime returns an Unimplemented error,
+// the client automatically falls back to unary RPCs.
+//
+// Deprecated: Use NewRemoteRuntimeServiceBuilder so that default values
+// (e.g. useStreaming) can be changed in a single place without updating
+// every callsite.
+func NewRemoteRuntimeService(ctx context.Context, endpoint string, connectionTimeout time.Duration, tp trace.TracerProvider, useStreaming bool) (internalapi.RuntimeService, error) {
+	return NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(connectionTimeout).
+		WithTracerProvider(tp).
+		WithUseStreaming(useStreaming).
+		Build(ctx)
 }
 
 // Close will shutdown the internal gRPC client connection.

--- a/staging/src/k8s.io/cri-client/pkg/remote_runtime_test.go
+++ b/staging/src/k8s.io/cri-client/pkg/remote_runtime_test.go
@@ -20,17 +20,19 @@ import (
 	"context"
 	"os"
 	"runtime"
+	"sync/atomic"
 	"testing"
 	"time"
 
 	sdktrace "go.opentelemetry.io/otel/sdk/trace"
 	"go.opentelemetry.io/otel/sdk/trace/tracetest"
-	oteltrace "go.opentelemetry.io/otel/trace"
 	"go.opentelemetry.io/otel/trace/noop"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
-	internalapi "k8s.io/cri-api/pkg/apis"
+	kubeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
 	apitest "k8s.io/cri-api/pkg/apis/testing"
 	fakeremote "k8s.io/cri-client/pkg/fake"
 	"k8s.io/cri-client/pkg/util"
@@ -51,21 +53,6 @@ func createAndStartFakeRemoteRuntime(t *testing.T) (*fakeremote.RemoteRuntime, s
 	return fakeRuntime, endpoint
 }
 
-func createRemoteRuntimeService(ctx context.Context, endpoint string, t *testing.T) internalapi.RuntimeService {
-	runtimeService, err := NewRemoteRuntimeService(ctx, endpoint, defaultConnectionTimeout, noop.NewTracerProvider(), false)
-
-	require.NoError(t, err)
-
-	return runtimeService
-}
-
-func createRemoteRuntimeServiceWithTracerProvider(ctx context.Context, endpoint string, tp oteltrace.TracerProvider, t *testing.T) internalapi.RuntimeService {
-	runtimeService, err := NewRemoteRuntimeService(ctx, endpoint, defaultConnectionTimeout, tp, false)
-	require.NoError(t, err)
-
-	return runtimeService
-}
-
 func TestGetSpans(t *testing.T) {
 	fakeRuntime, endpoint := createAndStartFakeRemoteRuntime(t)
 	defer func() {
@@ -82,8 +69,13 @@ func TestGetSpans(t *testing.T) {
 		sdktrace.WithBatcher(exp),
 	)
 	ctx := context.Background()
-	rtSvc := createRemoteRuntimeServiceWithTracerProvider(ctx, endpoint, tp, t)
-	_, err := rtSvc.Version(ctx, apitest.FakeVersion)
+	rtSvc, err := NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(defaultConnectionTimeout).
+		WithTracerProvider(tp).
+		Build(ctx)
+	require.NoError(t, err)
+	_, err = rtSvc.Version(ctx, apitest.FakeVersion)
 	require.NoError(t, err)
 	err = tp.ForceFlush(ctx)
 	require.NoError(t, err)
@@ -103,11 +95,127 @@ func TestVersion(t *testing.T) {
 	}()
 
 	ctx := context.Background()
-	rtSvc := createRemoteRuntimeService(ctx, endpoint, t)
+	rtSvc, err := NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(defaultConnectionTimeout).
+		Build(ctx)
+	require.NoError(t, err)
 	version, err := rtSvc.Version(ctx, apitest.FakeVersion)
 	require.NoError(t, err)
 	assert.Equal(t, apitest.FakeVersion, version.Version)
 	assert.Equal(t, apitest.FakeRuntimeName, version.RuntimeName)
+}
+
+// TestTracerProviderPropagation verifies that the otelgrpc stats handler
+// (which provides gRPC client context propagation) is installed when
+// WithTracerProvider is not called or is called with a non-nil provider, and
+// is skipped when WithTracerProvider(nil) is called explicitly.
+func TestTracerProviderPropagation(t *testing.T) {
+	exp := tracetest.NewInMemoryExporter()
+	tp := sdktrace.NewTracerProvider(sdktrace.WithBatcher(exp))
+	tracer := tp.Tracer("test")
+
+	cases := []struct {
+		name            string
+		withTP          func(*RemoteRuntimeServiceBuilder) *RemoteRuntimeServiceBuilder
+		wantTraceParent bool
+	}{
+		{
+			name:            "default propagates",
+			withTP:          func(b *RemoteRuntimeServiceBuilder) *RemoteRuntimeServiceBuilder { return b },
+			wantTraceParent: true,
+		},
+		{
+			name: "noop tracer provider propagates",
+			withTP: func(b *RemoteRuntimeServiceBuilder) *RemoteRuntimeServiceBuilder {
+				return b.WithTracerProvider(noop.NewTracerProvider())
+			},
+			wantTraceParent: true,
+		},
+		{
+			name:            "nil tracer provider disables propagation",
+			withTP:          func(b *RemoteRuntimeServiceBuilder) *RemoteRuntimeServiceBuilder { return b.WithTracerProvider(nil) },
+			wantTraceParent: false,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			endpoint, sawTraceParent, stop := startCapturingFakeRuntime(t)
+			defer stop()
+
+			ctx := context.Background()
+			b := NewRemoteRuntimeServiceBuilder().
+				WithEndpoint(endpoint).
+				WithConnectionTimeout(defaultConnectionTimeout)
+			rtSvc, err := tc.withTP(b).Build(ctx)
+			require.NoError(t, err)
+
+			rpcCtx, span := tracer.Start(ctx, "outer")
+			_, err = rtSvc.Version(rpcCtx, apitest.FakeVersion)
+			require.NoError(t, err)
+			span.End()
+
+			assert.Equal(t, tc.wantTraceParent, sawTraceParent.Load(), "traceparent header presence")
+		})
+	}
+}
+
+// startCapturingFakeRuntime starts a fake CRI runtime on a fresh endpoint and
+// returns a flag that the unary server interceptor flips to true the first
+// time it sees a "traceparent" header (the W3C tracecontext header the
+// otelgrpc client propagator injects) in incoming gRPC metadata.
+func startCapturingFakeRuntime(t *testing.T) (string, *atomic.Bool, func()) {
+	endpoint, err := fakeremote.GenerateEndpoint()
+	require.NoError(t, err)
+
+	var sawTraceParent atomic.Bool
+	interceptor := func(ctx context.Context, req any, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+		if md, ok := metadata.FromIncomingContext(ctx); ok && len(md.Get("traceparent")) > 0 {
+			sawTraceParent.Store(true)
+		}
+		return handler(ctx, req)
+	}
+
+	server := grpc.NewServer(grpc.UnaryInterceptor(interceptor))
+	fake := fakeremote.NewFakeRemoteRuntime()
+	kubeapi.RegisterRuntimeServiceServer(server, fake)
+	kubeapi.RegisterImageServiceServer(server, fake)
+	fake.RuntimeService.FakeStatus = &kubeapi.RuntimeStatus{
+		Conditions: []*kubeapi.RuntimeCondition{
+			{Type: kubeapi.RuntimeReady, Status: true},
+			{Type: kubeapi.NetworkReady, Status: true},
+		},
+	}
+
+	lis, err := util.CreateListener(endpoint)
+	require.NoError(t, err)
+	go func() { _ = server.Serve(lis) }()
+
+	stop := func() {
+		server.Stop()
+		if addr, _, err := util.GetAddressAndDialer(endpoint); err == nil {
+			if _, err := os.Stat(addr); err == nil {
+				if err := os.Remove(addr); err != nil {
+					t.Errorf("remove %q: %v", addr, err)
+				}
+			}
+		}
+	}
+	return endpoint, &sawTraceParent, stop
+}
+
+func TestBuildValidatesRequiredOptions(t *testing.T) {
+	ctx := context.Background()
+	_, err := NewRemoteRuntimeServiceBuilder().
+		WithConnectionTimeout(defaultConnectionTimeout).
+		Build(ctx)
+	require.ErrorContains(t, err, "endpoint is required")
+
+	_, err = NewRemoteRuntimeServiceBuilder().
+		WithEndpoint("unix:///tmp/cri-client-test.sock").
+		Build(ctx)
+	assert.ErrorContains(t, err, "connectionTimeout must be positive")
 }
 
 func TestNewRemoteRuntimeServiceUnixSocketEndpoint(t *testing.T) {
@@ -129,7 +237,11 @@ func TestNewRemoteRuntimeServiceUnixSocketEndpoint(t *testing.T) {
 	}()
 
 	ctx := context.Background()
-	rtSvc := createRemoteRuntimeService(ctx, endpoint, t)
+	rtSvc, err := NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(endpoint).
+		WithConnectionTimeout(defaultConnectionTimeout).
+		Build(ctx)
+	require.NoError(t, err)
 	version, err := rtSvc.Version(ctx, apitest.FakeVersion)
 	require.NoError(t, err)
 	assert.Equal(t, apitest.FakeVersion, version.Version)

--- a/test/e2e_node/util.go
+++ b/test/e2e_node/util.go
@@ -35,8 +35,6 @@ import (
 	"k8s.io/kubernetes/pkg/util/procfs"
 	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
 
-	"go.opentelemetry.io/otel/trace/noop"
-
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/runtime"
@@ -298,7 +296,11 @@ func getCRIClient(ctx context.Context) (internalapi.RuntimeService, internalapi.
 	const connectionTimeout = 2 * time.Minute
 	runtimeEndpoint := framework.TestContext.ContainerRuntimeEndpoint
 	useStreaming := utilfeature.DefaultFeatureGate.Enabled(features.CRIListStreaming)
-	r, err := remote.NewRemoteRuntimeService(ctx, runtimeEndpoint, connectionTimeout, noop.NewTracerProvider(), useStreaming)
+	r, err := remote.NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(runtimeEndpoint).
+		WithConnectionTimeout(connectionTimeout).
+		WithUseStreaming(useStreaming).
+		Build(ctx)
 	if err != nil {
 		return nil, nil, err
 	}
@@ -308,7 +310,11 @@ func getCRIClient(ctx context.Context) (internalapi.RuntimeService, internalapi.
 		//explicitly specified
 		imageManagerEndpoint = framework.TestContext.ImageServiceEndpoint
 	}
-	i, err := remote.NewRemoteImageService(ctx, imageManagerEndpoint, connectionTimeout, noop.NewTracerProvider(), useStreaming)
+	i, err := remote.NewRemoteImageServiceBuilder().
+		WithEndpoint(imageManagerEndpoint).
+		WithConnectionTimeout(connectionTimeout).
+		WithUseStreaming(useStreaming).
+		Build(ctx)
 	if err != nil {
 		return nil, nil, err
 	}

--- a/test/e2e_node_windows/util.go
+++ b/test/e2e_node_windows/util.go
@@ -23,8 +23,6 @@ import (
 	"flag"
 	"time"
 
-	"go.opentelemetry.io/otel/trace/noop"
-
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	internalapi "k8s.io/cri-api/pkg/apis"
 	remote "k8s.io/cri-client/pkg"
@@ -49,7 +47,11 @@ func getCRIClient(ctx context.Context) (internalapi.RuntimeService, internalapi.
 	const connectionTimeout = 2 * time.Minute
 	runtimeEndpoint := framework.TestContext.ContainerRuntimeEndpoint
 	useStreaming := utilfeature.DefaultFeatureGate.Enabled(features.CRIListStreaming)
-	r, err := remote.NewRemoteRuntimeService(ctx, runtimeEndpoint, connectionTimeout, noop.NewTracerProvider(), useStreaming)
+	r, err := remote.NewRemoteRuntimeServiceBuilder().
+		WithEndpoint(runtimeEndpoint).
+		WithConnectionTimeout(connectionTimeout).
+		WithUseStreaming(useStreaming).
+		Build(ctx)
 	if err != nil {
 		return nil, nil, err
 	}
@@ -59,7 +61,11 @@ func getCRIClient(ctx context.Context) (internalapi.RuntimeService, internalapi.
 		// explicitly specified.
 		imageManagerEndpoint = framework.TestContext.ImageServiceEndpoint
 	}
-	i, err := remote.NewRemoteImageService(ctx, imageManagerEndpoint, connectionTimeout, noop.NewTracerProvider(), useStreaming)
+	i, err := remote.NewRemoteImageServiceBuilder().
+		WithEndpoint(imageManagerEndpoint).
+		WithConnectionTimeout(connectionTimeout).
+		WithUseStreaming(useStreaming).
+		Build(ctx)
 	if err != nil {
 		return nil, nil, err
 	}

